src/Package.php

```diff
@@ -371,31 +372,43 @@ public function build(): Package
     public function boot(Module ...$defaultModules): bool
     {
         try {
+            // When package is done, nothing should happen to it calling boot again, but we call
+            // false to signal something is off.
+            if ($this->statusIs(self::STATUS_DONE)) {
```

Comment:
In case of an error we have `self::ACTION_FAILED_BOOT` set als alternative to `self::STATUS_DONE`.

a) Should we align naming here? (Breaking change)
b) Should we check that as well here?

Reply:
I do not understand.

`ACTION_FAILED_BOOT` is an _action_ hook, not a status.
This action is fired when an error occurs during the process inside `$package->boot()`.

So when `$package->boot()` is called, the following action might be called:

- `ACTION_BOOTED`
- `ACTION_FAILED_BOOT`.

Regarding _statuses_, `$package->boot()` is called, the following statuses might happen:

- `STATUS_BOOTING`
- `STATUS_BOOTED`
- `STATUS_DONE`
- `STATUS_FAILED`

The `STATUS_FAILED` might be after `STATUS_BOOTING` or after `STATUS_BOOTED`, but not after `STATUS_DONE`.

So, when `$package->boot()` is called, there are 3 possible scenarios:

**Case 1**

- Status is `STATUS_BOOTING`
- Status is `STATUS_BOOTED`
- Action `ACTION_BOOTED` is fired
- Status is `STATUS_DONE`

**Case 2**

- Status is `STATUS_BOOTING`
- Status is `STATUS_FAILED`
- Action `ACTION_FAILED_BOOT` is fired

**Case 3**

- Status is `STATUS_BOOTING`
- Status is `STATUS_BOOTED`
- Action `ACTION_BOOTED` is fired
- Status is `STATUS_FAILED`
- Action `ACTION_FAILED_BOOT` is fired

Note: `STATUS_FAILED` might also happen when `$package->build()` is called. In that case, even if the final status is the same, the action fired will be `ACTION_FAILED_BUILD`.

Finally, the `STATUS_FAILED` might also be reached in the case of a failure of `$package->connect()`. In that case, the action fired will be `ACTION_FAILED_CONNECTION`.

The names of all the statuses and actions look consistent to me.

To be noted: when a failure happens, if debug is on, after the `ACTION_FAILED_*` action is fired, exceptions are thrown, so nothing else can be done. Which means the status is not that relevant.

---

Regarding the `$this->statusIs(self::STATUS_DONE)`, that check allows executing without failures a thing like:

```php
$package->boot()->boot()->boot();
```

Of course, we don't expect such code, but multiple external packages might call `boot()` ending up in something similar.

If the first call to `->boot()` succeeds, the other calls do nothing because of the `if` check.

In the case the first call to `->boot()` fails, the status is `STATUS_FAILED`, and two things might happen:

- If debug is on, an exception is thrown, so other calls never get to run
- If debug is off, because the status is `STATUS_FAILED`, the second call goes into failure mode, but nothing really happens because the status is already failed. Even the `ACTION_FAILED_BOOT` is not fired again because there's [a check to run it once](https://github.com/inpsyde/modularity/blob/master/src/Package.php#L692).

In substance, multiple calls to `boot()` are uneventful regardless of whether the first call succeeds or fails.

Reply:
Sorry, seems like I've updated my comment while you were already writing the response. I had `ACTION_FAILED` in my original message instead of `STATUS_FAILED` which caused the confusion. 🙈 

What I meant was, if we really want to call complete `boot()` again when the Package is in `STATUS_FAILED` or if we want to return early to avoid throwing again (or maybe even successfully booting on next try)?

Reply:
@Chrico as written above, the method will return early, not just _that_ early.

Righ after that `if`, there's a call to [`->doBuild()`](https://github.com/inpsyde/modularity/blob/ab5c8a3715a1f3287811d1b253022a0682f9e8fe/src/Package.php#L390).

That method will [return pretty quickly if the status is failed](https://github.com/inpsyde/modularity/blob/ab5c8a3715a1f3287811d1b253022a0682f9e8fe/src/Package.php#L450-L454).

That means the code goes quickly to [`->assertStatus()`](https://github.com/inpsyde/modularity/blob/ab5c8a3715a1f3287811d1b253022a0682f9e8fe/src/Package.php#L394) which throws.

That exception is caught, and goes to [`->handleFailure()`](https://github.com/inpsyde/modularity/blob/ab5c8a3715a1f3287811d1b253022a0682f9e8fe/src/Package.php#L407-L408).

In [that method](https://github.com/inpsyde/modularity/blob/ab5c8a3715a1f3287811d1b253022a0682f9e8fe/src/Package.php#L774-L782), two things might happen, depending on the previous context:

- if the status is failed because of a previous `->boot()` call (something like `->boot()->boot()`) not much will happen. The only things that happens is that the `$this->lastError` is updated. Still, this is good for debug.
- if the status is failed because OF something else, then before `$this->lastError` is updated, the action `ACTION_FAILED_BOOT` is fired, passing the exception as argument.

This second case (which is probably more common than the first) allow us to have better logging/debugging.

Imagine a quite common code like the following:

```php
$package->addModule($x)->addModule($y)->boot();
```

And imagine that one the two `>addModule()` fails for any reason. If debug is off, so no exception is thrown when `>addModule()` fails, `->boot()` will be called, and we are in the second case from above. So we are going to emit the action `ACTION_FAILED_BOOT`.

Because the failed  `>addModule()` ended up firing `ACTION_FAILED_ADD_MODULE` it means that we will have a sequence of two actions:

- `ACTION_FAILED_ADD_MODULE`
- `ACTION_FAILED_BOOT`

If we setup a log system to listen to these actions, when we open the logs we would read logs like:

```
Failed add Module X - Previous: method X::y 1st argument expect int string given
Failed boot application - Previous: Failed adding module X
```

That very clearly describes what happened. The "previous" bit comes from the `$last->error` used as previous for the exceptions.

If we would exit early in `boot()` when the status is failed we would not get the second log entry, but I think it is good to have.

Reply:
Let's keep it as it is. Was also more an open discussion point from my end to clarify how we want to handle that. Should be good to go 👍🏻